components/script/dom/servoparser/mod.rs

```diff
@@ -1381,6 +1393,85 @@ impl TreeSink for Sink {
         let node = DomRoot::from_ref(&**node);
         vtable_for(&node).pop();
     }
+
+    fn allow_declarative_shadow_roots(&self, intended_parent: &Dom<Node>) -> bool {
+        intended_parent.owner_doc().allow_declarative_shadow_roots()
+    }
+
+    // <https://html.spec.whatwg.org/multipage/#parsing-main-inhead>
+    // A start tag whose tag name is "template"
+    // Attach shadow path
+    fn attach_declarative_shadow(
+        &self,
+        host: &Dom<Node>,
+        template: &Dom<Node>,
+        attrs: Vec<Attribute>,
+    ) -> Result<(), String> {
+        let host_element = host.downcast::<Element>().unwrap();
+
+        if host_element.shadow_root().is_some() {
+            return Err(String::from("Already in a shadow host"));
+        }
+
+        let template_element = template.downcast::<HTMLTemplateElement>().unwrap();
+
+        // Step 3-4-5-6.
```

Comment:
Please add spec text for these steps, even if it's a bit verbose.

The spec steps shift over time, so the numbers alone will likely refer to different steps in a couple of years.